tests/farmer_harvester/third_party_harvesters/test_third_party_harvesters.py

```diff
@@ -0,0 +1,331 @@
+        nonlocal validated_foliage_transaction
+        nonlocal validated_cc_vdf
+        nonlocal validated_rc_vdf
+        nonlocal validated_sub_slot_cc
+        nonlocal validated_sub_slot_rc
+        nonlocal finished_validating_data
+
+        assert request.message_data is not None
+        assert len(request.messages) > 0
+        assert len(request.messages) == len(request.message_data)
+
+        for i in range(len(request.messages)):
+            hash = request.messages[i]
+            src = request.message_data[i]
+            assert src
+
+            data: Optional[Streamable] = None
+            if src.kind == uint8(SigningDataKind.FOLIAGE_BLOCK_DATA):
+                data = FoliageBlockData.from_bytes(src.data)
+                assert (
+                    data.farmer_reward_puzzle_hash == farmer_reward_address
+                    or data.farmer_reward_puzzle_hash
+                    == bytes32(full_node.constants.GENESIS_PRE_FARM_FARMER_PUZZLE_HASH)
+                )
+                if data.farmer_reward_puzzle_hash == farmer_reward_address:
+                    validated_foliage_data = True
+            elif src.kind == uint8(SigningDataKind.FOLIAGE_TRANSACTION_BLOCK):
+                data = FoliageTransactionBlock.from_bytes(src.data)
+                validated_foliage_transaction = True
+            elif src.kind == uint8(SigningDataKind.CHALLENGE_CHAIN_VDF):
+                data = ClassgroupElement.from_bytes(src.data)
+                validated_cc_vdf = True
+            elif src.kind == uint8(SigningDataKind.REWARD_CHAIN_VDF):
+                data = ClassgroupElement.from_bytes(src.data)
+                validated_rc_vdf = True
+            elif src.kind == uint8(SigningDataKind.CHALLENGE_CHAIN_SUB_SLOT):
+                data = ChallengeChainSubSlot.from_bytes(src.data)
+                validated_sub_slot_cc = True
+            elif src.kind == uint8(SigningDataKind.REWARD_CHAIN_SUB_SLOT):
+                data = RewardChainSubSlot.from_bytes(src.data)
+                validated_sub_slot_rc = True
+            elif src.kind == uint8(SigningDataKind.PARTIAL):
+                # #NOTE: This data type is difficult to trigger, so it is
+                #        not tested for the time being.
+                # data = PostPartialPayload.from_bytes(src.data)
+                # validated_partial = True
+                pass
+
+            finished_validating_data = (
+                validated_foliage_data
+                and validated_foliage_transaction
+                and validated_cc_vdf
+                and validated_rc_vdf
+                and validated_sub_slot_cc
+                and validated_sub_slot_rc
+            )
+
+            assert data is not None
+            data_hash = data.get_hash()
+            assert data_hash == hash
+
+    async def intercept_farmer_new_proof_of_space(*args: Any) -> None:
+        nonlocal farmer
+        nonlocal farmer_reward_address
```

Comment:
I don't understand the rationale behind declaring these as `nonlocal`. the outer scope here is the global scope, isn't it?